### ISSUE TITLE
Filtered changes are accumulated until the build is triggered.

### ISSUE THREAD

**Issue by chrwen-omicron** [buildbot/buildbot#3189]
I use a change filter to ignore certain parts of our monorepo, but those ignored changes and the responsible users are only accumulated until the next build.
I thought this is only the case with fileIsImportant and onlyImportant set to False.

**Comment by seankelly** [comment]
Not sure what the issue is. The ChangeFilter will filter out changes entirely and they won't show up when the build happens. "Unimportant" changes will appear unless `onlyImportant` is `True`.

The docs say this about `onlyImportant`:
> A boolean that, when `True`, only adds important changes to the buildset as specified in the `fileIsImportant` callable. This means that unimportant changes are ignored the same way a `change_filter` filters changes. This defaults to `False` and only applies when `fileIsImportant` is given.

**Comment by chrwen-omicron** [comment]
Exactly that's currently not the case for me. Previously filtered/ignored changes appear in the next build, my ChangeFilter behaves just like the documented combination of fileIsImportant and onlyImportant=False.
I have changes in my builds from hours ago.

**Comment by tardyp** [comment]
Ignored changes are ignored for the matter of the creation of the
buildrequests. Once one important change triggers a build those changes are
effectively in that build hence they are shown in the build UI.

That said, the algorithm which computes the list of changes for a build is
too simplistic. It could use some improvements.

Le mar. 16 mai 2017 09:47, Christian Wenk <notifications@github.com> a
écrit :

**Comment by chrwen-omicron** [comment]
But what's the difference then between "accumulated" and ignored. The documentation states
> ... only adds important changes to the buildset ...
I have the problem that there are responsible users in a build who are in fact not responsible. Is there a way to really ignore a change?

**Comment by tardyp** [comment]
Indeed, importantChanges are completely ignored for responsibility list generation.
I am afraid you will need to create your own reponsibility list generation for your usecase :-(

**Comment by chrwen-omicron** [comment]
So what does the onlyImportant flag do if the changes are accumulated anyway?

**Comment by tardyp** [comment]
It prevents creation of unimportant builds

**Comment by chrwen-omicron** [comment]
That's what the fileIsImportant callable is doing, according to the documentation onlyImportant decides whether changes are accumulated or ignored like ChangeFilter. But in the end using a ChangeFilter also accumulates changes until the next build ...
I'm sorry, I just try get the difference.

**Comment by tardyp** [comment]
I think for this level of detail, the truth is the source code. There is a strong possibility that this particular sentence of the doc has been made obsolete by some change done since buildbot nine fork.

A big difference between nine and eight is that in eight, a sourcestamp was attached a list of changes.
Now the source stamp only points to one change, the change that has the same ssid. This decision was made at the very beginning of nine branching by @djmitche . I am not sure exactly what was the reasoning behind that. I guess it was to simplify the data model and improve efficiency. see
1a80c3cfccdf066dbab9331723e4fcba07066366 and a3fcc91ee3982eac7bd9d78a5c5202c0a9602986

The important value is still kept in the scheduler_change table https://github.com/buildbot/buildbot/blob/master/master/buildbot/db/model.py#L502

A solution to this problem would be to modify getChangesForBuild and take this information in account. We would need to join with the buildrequest and buildset table to get the schedulerid, so that we can fetch the information whether that change was important or not for that build.
https://github.com/buildbot/buildbot/blob/master/master/buildbot/db/changes.py#L201

**Comment by djmitche** [comment]
The idea was that, given two sourcestamps, it would be possible to find all changes "between" those stamps and display them.  However, that is too late to build a blamelist, because as you've noted, whether a change is important is specific to the scheduler.

It might make most sense for the scheduler to generate the blamelist when it creates the build request?

**Comment by chrwen-omicron** [comment]
But in case of a ChangeFilter the important value is not tracked https://github.com/buildbot/buildbot/blob/master/master/buildbot/schedulers/base.py#L198
